assets/scripts/segments/resize_drag.js

```diff
@@ -0,0 +1,97 @@
+import { setIgnoreStreetChanges } from '../streets/data_model'
+import { INFO_BUBBLE_TYPE_SEGMENT } from '../info_bubble/constants'
+import { infoBubble } from '../info_bubble/info_bubble'
+import {
+  RESIZE_TYPE_INITIAL,
+  RESIZE_TYPE_PRECISE_DRAGGING,
+  resizeSegment,
+  cancelFadeoutControls
+} from './resizing'
+import { segmentsChanged, getSegmentEl } from './view'
+import { TILE_SIZE } from './constants'
+import store from '../store'
+
+let originalWidth
+
+function getActiveSegmentWidth () {
+  const activeSegment = store.getState().ui.activeSegment
+  const segment = store.getState().street.segments[activeSegment]
+  return segment.width
+}
+
+export function beginSegmentResize () {
+  originalWidth = getActiveSegmentWidth()
+
+  setIgnoreStreetChanges(true)
+  document.body.classList.add('segment-resize-dragging')
+
+  infoBubble.hide()
+  cancelFadeoutControls()
+}
+
+export function duringSegmentResize (delta) {
+  const previewWidth = originalWidth + (delta / TILE_SIZE * 2)
+  const activeSegment = store.getState().ui.activeSegment
+
+  // TODO: put back "precise" once we know how to get shift key.
+  // const precise = event.shiftKey
+  // let resizeType
+  // if (precise) {
+  //   resizeType = RESIZE_TYPE_PRECISE_DRAGGING
+  // } else {
+  //   resizeType = RESIZE_TYPE_DRAGGING
+  // }
+  const resizeType = RESIZE_TYPE_PRECISE_DRAGGING
+
+  resizeSegment(activeSegment, resizeType, previewWidth)
+}
+
+export function endSegmentResize () {
+  const activeSegment = store.getState().ui.activeSegment
+  handleSegmentResizeEnd(activeSegment)
+}
+
+// TODO: This is no longer used anywhere (the keydown button that used to call this is no longer
+// set), but we should consider making it possible to cancel resizing again.
+export function cancelSegmentResize () {
+  const activeSegment = store.getState().ui.activeSegment
+
+  // Resize segment back to original width
+  resizeSegment(activeSegment, RESIZE_TYPE_INITIAL, originalWidth, true, false)
+
+  handleSegmentResizeEnd(activeSegment)
+}
+
+// TODO: use suppressMouseEnter on <StreetEditable /> state
+let _suppressMouseEnter = false
+
+export function suppressMouseEnter () {
+  return _suppressMouseEnter
+}
+
+function handleSegmentResizeEnd (activeSegment) {
+  setIgnoreStreetChanges(false)
+
+  // If width has changed changed, register this
+  if (getActiveSegmentWidth() === originalWidth) {
```

Comment:
Should this be `getActiveSegmentWidth() !== originalWidth)` since you're checking if the width has changed? Doesn't the last call to `duringSegmentResize` also call `segmentsChanged'? Sorry I don't think I understand what this check is for.

Reply:
Yep, wrong comparator.

As for the check: While resizing, we tell Streetmix to "ignore street changes" which means that the changes are in a "preview" mode, and it won't save the intermediate states to the server. At the end of the resize action, we turn off `setIgnoreStreetChanges` then we call `segmentsChanged()` again to save the street. You're right that it's already being called in `duringSegmentResize()` and it does extraneous work here. The important part is the call to `saveStreetToServerIfNecessary()` within `segmentsChanged()` so I'll move that up to here.